from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from routers import books
from contextlib import asynccontextmanager
from datetime import datetime
import time

@asynccontextmanager
async def lifespan(app: FastAPI):
    # –ö–æ–¥ –ø—Ä–∏ –∑–∞–ø—É—Å–∫–µ
    with open("state.txt", "a", encoding="utf-8") as f:
        f.write(f"üöÄ START at {datetime.now()}\n")

    yield  # ‚¨ÖÔ∏è –ü–µ—Ä–µ—Ö–æ–¥ –∫ –æ–±—Ä–∞–±–æ—Ç–∫–µ –∑–∞–ø—Ä–æ—Å–æ–≤

    # –ö–æ–¥ –ø—Ä–∏ –∑–∞–≤–µ—Ä—à–µ–Ω–∏–∏
    with open("state.txt", "a", encoding="utf-8") as f:
        f.write(f"üõë SHUTDOWN at {datetime.now()}\n")

app = FastAPI(lifespan=lifespan)

@app.middleware("http")
async def log_request(request: Request, call_next):
    start_time = time.time()

    client_ip = request.client.host
    method = request.method
    url = str(request.url)

    print(f"[{method}] {url} ‚Äî –æ—Ç {client_ip}")

    response = await call_next(request)

    duration = time.time() - start_time
    print(f"–û–±—Ä–∞–±–æ—Ç–∞–Ω –∑–∞ {duration:.3f} —Å–µ–∫")

    return response

app.include_router(books.router)

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request: Request, exc: RequestValidationError):
    return JSONResponse(
        status_code=422,
        content={
            "error": "–û—à–∏–±–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏",
            "details": exc.errors()
        }
    )
